### PR TITLE
[3.9] Log the from version on actionlogs when Joomla update

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #30499 .

### Summary of Changes
Log the from version on  Joomla update event on the Actionlogs

### Testing Instructions
apply pr
joomlaupdate -> options
change to Custom URL
and put this manifest https://ci.joomla.org/artifacts/joomla/joomla-cms/staging/30714/downloads/35635/pr_list.xml
is the one available on bottom  near Download — Prebuilt packages are available for download.

### Actual result BEFORE applying this Pull Request
information about from version not available

### Expected result AFTER applying this Pull Request
information about from version  available
image

### CODE REVIEW

File: administrator/components/com_joomlaupdate/models/default.php [hunk 1]
```diff
@@ -18,7 +18,7 @@
  * @since  2.5.4
  */
 class JoomlaupdateModelDefault extends JModelLegacy
-{
+{
```
Comment (richard67) [inline administrator/components/com_joomlaupdate/models/default.php]:
```suggestion
{
```

File: administrator/components/com_joomlaupdate/models/default.php [hunk 2]
```diff
@@ -923,8 +923,10 @@ public function cleanUp()
 		// Unset the update filename from the session.
 		JFactory::getApplication()->setUserState('com_joomlaupdate.file', null);
 
+		$oldVersion = JFactory::getApplication()->getUserState('com_joomlaupdate.oldversion');
 		// Trigger event after joomla update.
```
Comment (richard67) [inline administrator/components/com_joomlaupdate/models/default.php:927]:
```suggestion

		// Trigger event after joomla update.
```

File: plugins/actionlog/joomla/joomla.php [hunk 3]
```diff
@@ -1053,7 +1053,7 @@ public function onAfterPurge($group = 'all')
 	 *
```
Comment (richard67) [inline plugins/actionlog/joomla/joomla.php:1055]:
```suggestion
	 * @param   string  $oldVersion  The Joomla version before the update
	 *
	 * @return  void
	 *
```

Comment (richard67) [reply plugins/actionlog/joomla/joomla.php:1055]: This and the 2 previous suggested changes will fix the PHPCS errors reported by Drone here: [https://ci.joomla.org/joomla/joomla-cms/35550/1/3](https://ci.joomla.org/joomla/joomla-cms/35550/1/3).

File: plugins/actionlog/joomla/joomla.php [hunk 4]
```diff
@@ -1050,10 +1050,13 @@ public function onAfterPurge($group = 'all')
 	 * Method is called after user update the CMS.
 	 *
 	 * @return  void
+	 * @param   string  $oldVersion  The Joomla version before the update
+	 *
+	 * @return  void
```
Comment (richard67) [inline plugins/actionlog/joomla/joomla.php]:
```suggestion
	 * @param   string  $oldVersion  The Joomla version before the update
	 *
	 * @return  void
```
Ah, my mistake with previous suggestion. Should be only one return value doc comment.

**Comment by richard67** [comment]
@alikon It seems that it needs to apply the patch of this PR before doing the update to that package which you have linked.

**Comment by alikon** [comment]
Yes i forgot that, I'll add it

Il lun 21 set 2020, 16:05 Richard Fath <notifications@github.com> ha
scritto:

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on c3c84807c18ed361a96ec36ce3936b41628ab386<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30714">issues.joomla.org/tracker/joomla-cms/30714</a>.</sub>

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on c3c84807c18ed361a96ec36ce3936b41628ab386<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30714">issues.joomla.org/tracker/joomla-cms/30714</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30714">issues.joomla.org/tracker/joomla-cms/30714</a>.</sub>

**Comment by SharkyKZ** [comment]
This could be a B/C break if `onJoomlaAfterUpdate` is triggered by 3rd party extension.

**Comment by HLeithner** [comment]
Why should onJoomlaAfterUpdate be triggered by a 3rd party extension? I mean we have no way to protect events but do you think a 3rd party extension triggers this?

Or do you mean the plugin should add a default value for the $oldVersion parameter?

**Comment by SharkyKZ** [comment]
Given that this event was recently added, is undocumented and has a very limited use case, I don't think anyone is actually using it. But if you want to be really careful, adding a default value would be a sensible thing to do.

**Comment by alikon** [comment]
did you mean someting like `public function onJoomlaAfterUpdate($oldVersion = 'unknown')
 ` ?

**Comment by richard67** [comment]
Maybe null or empty string as default, and in the code use a JTEXT for 'unknown' if the parameter is empty, so it can be translated?

On the other hand, that seems to be a bit over-engineered for something which should never happen.

So I'd be ok with anything, no default like now, or hard-coded like suggested above.

**Comment by HLeithner** [comment]
Yes please add ="unknown"

**Comment by alikon** [comment]
added

**Comment by HLeithner** [comment]
Thanks, plz 2 tests and can be merged

**Comment by HLeithner** [comment]
plz also with a plugin that triggers the event without parameter.

**Comment by richard67** [comment]
Wouldn't is be better to use an empty string as default value for `$oldVersion` and use `JText::_('JLIB_UNKNOWN')` if `$oldVersion` is empty?

**Comment by richard67** [comment]
@HLeithner I don't have such plugin so I can't test.

**Comment by SharkyKZ** [comment]
>Wouldn't is be better to use an empty string as default value for $oldVersion and use JText::_('JLIB_UNKNOWN') if $oldVersion is empty?

IMO should be `null`. But realistically it makes no difference. It's only used internally by this plugin.

**Comment by richard67** [comment]
@HLeithner Shouldn't a code review be sufficient for the default value part of this PR? Does it really need to write a plugin to test this?

**Comment by HLeithner** [comment]
if you use null (without language) you will end in a broken text string "...updated Joomla from to 3.9.22"

Actually while testing you see this broken string, so I would suggest to catch this case better and change the code from

$oldVersion='unknown' to $oldVersion=null and test on if empty($oldVersion) and set the unkown version string...

@richard67 no you don't need to create a plugin you only have to trigger the event for example in a template:

`JFactory::getApplication()->triggerEvent('onJoomlaAfterUpdate');`

or
```
JFactory::getApplication()->triggerEvent('onJoomlaAfterUpdate', array(null));
JFactory::getApplication()->triggerEvent('onJoomlaAfterUpdate', array(''));
JFactory::getApplication()->triggerEvent('onJoomlaAfterUpdate', array('5.0.0));
```

**Comment by richard67** [comment]
@HLeithner 

> no you don't need to create a plugin

Well, you wrote "plz also with a plugin that triggers the event without parameter." 😛

**Comment by richard67** [comment]
@alikon Could you change it to

```
	public function onJoomlaAfterUpdate($oldVersion = null)
	{
...		
		if (empty($oldVersion))
		{			
			$oldVersion = JText::_('JLIB_UNKNOWN');	
		}
```
?

File: plugins/actionlog/joomla/joomla.php [hunk 5]
```diff
@@ -1055,12 +1055,12 @@ public function onAfterPurge($group = 'all')
...
 	 * @since   3.9.21
 	 */
-	public function onJoomlaAfterUpdate($oldVersion = 'unknown')
+	public function onJoomlaAfterUpdate($oldVersion = null)
 	{
 		$context = $this->app->input->get('option');
 		$user    = JFactory::getUser();
 		
-		if ($oldVersion === 'unknown')
+		if empty($oldVersion)
```
Comment (richard67) [inline plugins/actionlog/joomla/joomla.php]:
```suggestion
		if (empty($oldVersion))
```

Comment (richard67) [reply plugins/actionlog/joomla/joomla.php]: @alikon Please fix that.

**Comment by HLeithner** [comment]
> Well, you wrote "plz also with a plugin that triggers the event without parameter." 😛

wrong wording sorry

**Comment by richard67** [comment]
> wrong wording sorry

Now you say that? I almost have my plugin ready 🤣 . (joking)

**Comment by richard67** [comment]
I've updated the custom update URL in the testing instructions to the one of the new build.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on ee6b10f32d8a5f23e23a082ae92dbb1be57e16e3<br /><br />1. Real test as described:

Result: "User whoeveriamdoesnotmatter updated Joomla from 3.9.22-dev to 3.9.22-dev+pr.30714"

2. Following in index.php of the isis template:

```
$app->triggerEvent('onJoomlaAfterUpdate');
$app->triggerEvent('onJoomlaAfterUpdate', array(null));
$app->triggerEvent('onJoomlaAfterUpdate', array(''));
$app->triggerEvent('onJoomlaAfterUpdate', array('5.0.0'));
```

Result: 3 times "User whoeveriamdoesnotmatter updated Joomla from Unknown to 3.9.22-dev+pr.30714", one time "User whoeveriamdoesnotmatter updated Joomla from 5.0.0 to 3.9.22-dev+pr.30714"<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30714">issues.joomla.org/tracker/joomla-cms/30714</a>.</sub>

**Pull request merged**